ranger/core/actions.py

```diff
@@ -303,91 +322,161 @@ def get_macros(self):  # pylint: disable=too-many-branches,too-many-statements
-        else:
-            macros['t'] = MACRO_FAIL
+        macros['c'] = lambda: [
+            fl.path for fl in self.fm.copy_buffer] if self.fm.copy_buffer else MACRO_FAIL
 
-        if self.fm.thisdir:
-            macros['d'] = self.fm.thisdir.path
-        else:
-            macros['d'] = '.'
+        def get_tagged():
+            if self.fm.thisdir and self.fm.thisdir.files:
+                res = [fl.relative_path for fl in self.fm.thisdir.files
+                       if fl.realpath in self.fm.tags or []]
+            else:
+                res = MACRO_FAIL
+            return res
+        macros['t'] = get_tagged
+
+        macros['d'] = lambda: self.fm.thisdir.path if self.fm.thisdir else '.'
 
         # define d/f/p/s macros for each tab
-        for i in range(1, 10):
+
+        def get_tab_dir(i):
+            try:
+                tab = self.fm.tabs[i]
+            except KeyError:
+                return MACRO_FAIL
+
+            if not tab.thisdir:
+                return MACRO_FAIL
+
+            return tab.thisdir.path
+
+        def get_tab_path(i):
             try:
                 tab = self.fm.tabs[i]
             except KeyError:
-                continue
+                return MACRO_FAIL
+
+            if not tab.thisdir:
+                return MACRO_FAIL
+
             tabdir = tab.thisdir
-            if not tabdir:
-                continue
-            i = str(i)
-            macros[i + 'd'] = tabdir.path
-            if tabdir.get_selection():
-                macros[i + 'p'] = [os.path.join(tabdir.path, fl.relative_path)
-                                   for fl in tabdir.get_selection()]
-                macros[i + 's'] = [fl.path for fl in tabdir.get_selection()]
-            else:
-                macros[i + 'p'] = MACRO_FAIL
-                macros[i + 's'] = MACRO_FAIL
-            if tabdir.pointed_obj:
-                macros[i + 'f'] = tabdir.pointed_obj.path
-            else:
-                macros[i + 'f'] = MACRO_FAIL
+            return ([os.path.join(tabdir.path, fl.relative_path)
+                     for fl in tabdir.get_selection()]
+                    if tabdir.get_selection() else MACRO_FAIL)
+
+        def get_tab_selection(i):
+            try:
+                tab = self.fm.tabs[i]
+            except KeyError:
+                return MACRO_FAIL
+
+            if not tab.thisdir:
+                return MACRO_FAIL
+
+            tabdir = tab.thisdir
+            return ([fl.path for fl in tabdir.get_selection()]
+                    if tabdir.get_selection() else MACRO_FAIL)
+
+        def get_tab_file(i):
+            try:
+                tab = self.fm.tabs[i]
+            except KeyError:
+                return MACRO_FAIL
+
+            if not tab.thisdir:
+                return MACRO_FAIL
+
+            tabdir = tab.thisdir
+            return (tabdir.pointed_obj.path
+                    if tabdir.pointed_obj else MACRO_FAIL)
+
+        for i in range(1, 10):
```

Comment:
What do you mean exactly with `return lambda: whatever`. I Tried writing the macros as:
```
for i in range(1, 10):
            macros[str(i) + "d"] = lambda: get_tab_dir(i)
            ...
```
and this does not work. I think(I am not sure though) it will probably work if `get_tab_dir`  returns a function that takes an integer and return the dir of the corresponding tab. Is this what you are suggesting?

Reply:
No, I'm suggesting `get_tab_dir` returns a lambda.

Reply:
Ok. Would't that make the levels of function indirection bit too much too handle. The functions `get_tab_*' contain some multi line logic themselves so returning a labmda would require to create one more function in the corresponding function and return it. The dependency introduced by partial is with the functools package which is in the standard python library even for python 2.7, so is there actually a dependency? And the use of partial is pretty clear for anyone that knows what partial does so is there a problem there?

Reply:
I don't think one extra level's too much to handle. It's basically just manually doing what `partial` does. And I disagree it's obvious that `partial` on a single argument function returns an argumentless function, there's nothing partial about providing a function with all its arguments. Functools in general isn't all that pythonic, that's too bad but this is too obscure a use to justify it imo.

And since the bodies are multiline you can just return a nested function definition, we just need it to close over the argument.